src/group_theory/complement.lean

```diff
@@ -407,3 +408,73 @@ begin
 end
 
 end subgroup
+
+namespace subgroup
+
+open equiv function mem_left_transversals mul_action mul_action.quotient zmod
+
+universe u
+
+variables {G : Type u} [group G] (H : subgroup G) (g : G)
+
+/-- Partition `G ⧸ H` into orbits of the action of `g : G`. -/
+noncomputable def quotient_equiv_sigma_zmod : G ⧸ H ≃
+  Σ (q : quotient (orbit_rel (zpowers g) (G ⧸ H))), zmod (minimal_period ((•) g) q.out') :=
```

Comment:
I'd lean towards eliminating this `q.out'`, which I was able to do as:
```
noncomputable def orbit_rel_zpowers_period (q : quotient (orbit_rel (zpowers g) (G ⧸ H))) : ℕ :=
quotient.lift_on' q (minimal_period ((•) g)) $ λ a b (h : _ ∈ _), begin
  obtain ⟨⟨_, z, rfl⟩, rfl⟩ := h,
  dsimp [subgroup.smul_def],
  rw minimal_period_eq_minimal_period_iff,
  intro n,
  simp_rw [is_periodic_pt, is_fixed_pt],
  simp,
  rw [smul_smul, (((commute.refl g).pow_left n).zpow_right z).eq, mul_smul],
  rw smul_left_cancel_iff,
end

lemma orbit_rel_zpowers_period_mk (x : G ⧸ H) :
  orbit_rel_zpowers_period H g (quotient.mk' x) = minimal_period ((•) g) x := rfl
```
It's still noncomputable, but it has a nicer defeq than out.

The uglyness of that proof suggests that either I'm missing some API, or it's missing.

Reply:
I'll look into adding the appropriate API. However, I'm not so sure about eliminating this `quotient.out'`.

There is always going to be a `quotient.out'` coming from `self_equiv_sigma_orbits`, and keeping the `minimal_period` in terms of that `quotient.out'` makes all of these proofs a bit easier. Otherwise, every proof would involve rewriting `minimal_period ((•) g) q.out' = orbit_zpowers_period H g q`, and some nice defeqs would be lost.

Basically, the whole point of this definition is that you have to choose a representative from each orbit to correspond to `0 : zmod`, so I don't see a problem with mentioning that the minimal period of that representative, especially since future lemmas refer to `g ^ k • q.out'`, which mesh more nicely with `minimal_period ((•) g) q.out'` than `orbit_zpowers_period H g q`.

What do you think?

Reply:
> There is always going to be a `quotient.out'` coming from `self_equiv_sigma_orbits`,

#14591 adds a version without that `out'`.

Reply:
Ok, I'll try playing around with that version. I'm still worried about losing some nice defeqs, but we'll see.

Reply:
I played around with this a bit more (below), and to me it looks eliminating these `quotient.out'`s loses some nice defeqs and makes things more painful. In particular, `quotient_equiv_sigma_zmod_symm_apply` goes from defeq to painful (in fact, it's currently not clear to me how to prove it). The issue is that there is a `quotient.out'` coming from the choice of representative from each orbit corresponding to `0 : zmod`, and eliminating these other `quotient.out'`s makes things less compatible.

```
namespace subgroup

open equiv function mem_left_transversals mul_action mul_action.quotient zmod

variables {G : Type*} [group G] (g : G) (β : Type*) [mul_action G β]

noncomputable def orbit_zpowers_period (q : quotient (orbit_rel (zpowers g) β)) : ℕ :=
quotient.lift_on' q (minimal_period ((•) g)) $ λ a b (h : _ ∈ _), begin
  obtain ⟨⟨_, z, rfl⟩, rfl⟩ := h,
  dsimp [subgroup.smul_def],
  rw minimal_period_eq_minimal_period_iff,
  intro n,
  simp_rw [is_periodic_pt, is_fixed_pt],
  simp,
  rw [smul_smul, (((commute.refl g).pow_left n).zpow_right z).eq, mul_smul],
  rw smul_left_cancel_iff,
end

lemma orbit_zpowers_period_mk (x : β) :
  orbit_zpowers_period g β (quotient.mk' x) = minimal_period ((•) g) x := rfl

lemma minimal_period_out' (q : quotient (orbit_rel (zpowers g) β)) :
  minimal_period ((•) g) q.out' = orbit_zpowers_period g β q :=
by rw [←q.out_eq', orbit_zpowers_period_mk, q.out_eq']

end subgroup

namespace mul_action

variables (α β : Type*) [group α] [mul_action α β]

@[reducible]
def orbit_rel.quotient := (quotient $ orbit_rel α β)

variables {α β}

def orbit_rel.quotient.orbit (x : orbit_rel.quotient α β) : set β :=
quotient.lift_on' x (orbit α) $ λ _ _, mul_action.orbit_eq_iff.2

@[simp]
lemma orbit_rel.quotient.orbit_mk (b : β) :
  orbit_rel.quotient.orbit (quotient.mk' b : orbit_rel.quotient α β) = orbit α b := rfl

lemma orbit_rel.quotient.orbit_eq_orbit_out (x : orbit_rel.quotient α β)
  {φ : orbit_rel.quotient α β → β} (hφ : function.right_inverse φ quotient.mk') :
  orbit_rel.quotient.orbit x = orbit α (φ x) :=
begin
  conv_lhs { rw ←hφ x },
  induction x using quotient.induction_on',
  refl,
end

lemma orbit_rel.quotient.orbit_eq_orbit_out' (x : orbit_rel.quotient α β) :
  orbit_rel.quotient.orbit x = orbit α x.out' :=
orbit_rel.quotient.orbit_eq_orbit_out x quotient.out_eq'

variables (α) (β)

local notation `Ω` := orbit_rel.quotient α β

def my_self_equiv_sigma_orbits' : β ≃ Σ ω : Ω, ω.orbit :=
calc  β
    ≃ Σ (ω : Ω), {b // quotient.mk' b = ω} : (equiv.sigma_fiber_equiv quotient.mk').symm
... ≃ Σ (ω : Ω), ω.orbit :
        equiv.sigma_congr_right (λ ω, equiv.subtype_equiv_right $
          λ x, by {
            induction ω using quotient.induction_on',
            exact quotient.eq' })

def my_self_equiv_sigma_orbits : β ≃ Σ (ω : Ω), orbit α ω.out' :=
(my_self_equiv_sigma_orbits' α β).trans $ equiv.sigma_congr_right $ λ i,
  equiv.set.of_eq $ orbit_rel.quotient.orbit_eq_orbit_out _ quotient.out_eq'

noncomputable def my_orbit_zpowers_equiv (a : α) (q : orbit_rel.quotient (subgroup.zpowers a) β) :
  q.orbit ≃ zmod (subgroup.orbit_zpowers_period a β q) :=
begin
  rw [orbit_rel.quotient.orbit_eq_orbit_out', ←subgroup.minimal_period_out'],
  exact orbit_zpowers_equiv a q.out',
end

lemma my_orbit_zpowers_equiv_symm_apply (a : α) (q : quotient (orbit_rel (subgroup.zpowers a) β))
  (k : zmod (subgroup.orbit_zpowers_period a β q)) :
  (my_orbit_zpowers_equiv α β a q).symm k =
    ⟨a ^ (k : ℤ) • q.out', begin
      rw orbit_rel.quotient.orbit_eq_orbit_out',
      exact mem_orbit q.out' ((⟨a, subgroup.mem_zpowers a⟩ : subgroup.zpowers a) ^ (k : ℤ)),
    end⟩ :=
begin
  rw [subtype.ext_iff, subtype.coe_mk],
  sorry,
end

end mul_action

namespace subgroup

open equiv function mem_left_transversals mul_action mul_action.quotient zmod

universe u

variables {G : Type u} [group G] (H : subgroup G) (g : G)

/-- Partition `G ⧸ H` into orbits of the action of `g : G`. -/
noncomputable def quotient_equiv_sigma_zmod : G ⧸ H ≃
  Σ (q : quotient (orbit_rel (zpowers g) (G ⧸ H))), zmod (orbit_zpowers_period g (G ⧸ H) q) :=
(my_self_equiv_sigma_orbits' (zpowers g) (G ⧸ H)).trans
  (sigma_congr_right (λ q, my_orbit_zpowers_equiv G (G ⧸ H) g q))

lemma quotient_equiv_sigma_zmod_symm_apply
  (q : quotient (orbit_rel (zpowers g) (G ⧸ H))) (k : zmod (orbit_zpowers_period g (G ⧸ H) q)) :
  (quotient_equiv_sigma_zmod H g).symm ⟨q, k⟩ = g ^ (k : ℤ) • q.out' :=
begin
  simp only [quotient_equiv_sigma_zmod, symm_trans_apply, sigma_congr_right_symm,
    sigma_congr_right_apply, my_orbit_zpowers_equiv_symm_apply],
  refl,
end

lemma quotient_equiv_sigma_zmod_apply (q : quotient (orbit_rel (zpowers g) (G ⧸ H))) (k : ℤ) :
  quotient_equiv_sigma_zmod H g (g ^ k • q.out') = ⟨q, k⟩ :=
by rw [apply_eq_iff_eq_symm_apply, quotient_equiv_sigma_zmod_symm_apply,
  zmod.coe_int_cast, ←minimal_period_out', zpow_smul_mod_minimal_period]

/-- The transfer transversal as a function. -/
noncomputable def transfer_function : G ⧸ H → G :=
λ q, g ^ ((quotient_equiv_sigma_zmod H g q).2 : ℤ) * (quotient_equiv_sigma_zmod H g q).1.out'.out'

lemma transfer_function_apply (q : G ⧸ H) : transfer_function H g q =
  g ^ ((quotient_equiv_sigma_zmod H g q).2 : ℤ) * (quotient_equiv_sigma_zmod H g q).1.out'.out' :=
rfl

lemma coe_transfer_function (q : G ⧸ H) : ↑(transfer_function H g q) = q :=
by rw [transfer_function_apply, ←smul_eq_mul, coe_smul_out',
  ←quotient_equiv_sigma_zmod_symm_apply, sigma.eta, symm_apply_apply]

/-- The transfer transversal as a set. -/
def transfer_set : set G :=
set.range (transfer_function H g)

/-- The transfer transversal. -/
def transfer_transversal : left_transversals (H : set G) :=
⟨transfer_set H g, range_mem_left_transversals (coe_transfer_function H g)⟩

lemma transfer_transversal_apply (q : G ⧸ H) :
  ↑(to_equiv (transfer_transversal H g).2 q) = transfer_function H g q :=
to_equiv_apply (coe_transfer_function H g) q

lemma transfer_transversal_apply'
  (q : quotient (orbit_rel (zpowers g) (G ⧸ H))) (k : zmod (orbit_zpowers_period g (G ⧸ H) q)) :
  ↑(to_equiv (transfer_transversal H g).2 (g ^ (k : ℤ) • q.out')) = g ^ (k : ℤ) * q.out'.out' :=
by rw [transfer_transversal_apply, transfer_function_apply,
  ←quotient_equiv_sigma_zmod_symm_apply, apply_symm_apply]

lemma transfer_transversal_apply''
  (q : quotient (orbit_rel (zpowers g) (G ⧸ H))) (k : zmod (orbit_zpowers_period g (G ⧸ H) q)) :
  ↑(to_equiv (g • transfer_transversal H g).2 (g ^ (k : ℤ) • q.out')) =
    if k = 0 then g ^ minimal_period ((•) g) q.out' * q.out'.out' else g ^ (k : ℤ) * q.out'.out' :=
begin
  rw [smul_apply_eq_smul_apply_inv_smul, transfer_transversal_apply, transfer_function_apply,
      ←mul_smul, ←zpow_neg_one, ←zpow_add, quotient_equiv_sigma_zmod_apply, smul_eq_mul,
      ←mul_assoc, ←zpow_one_add, int.cast_add, int.cast_neg, int.cast_one, int_cast_cast,
      cast_id', id.def, ←sub_eq_neg_add, cast_sub_one, add_sub_cancel'_right],
  by_cases hk : k = 0,
  { rw [if_pos hk, if_pos hk, int.nat_cast_eq_coe_nat, zpow_coe_nat, minimal_period_out'] },
  { rw [if_neg hk, if_neg hk] },
end

end subgroup
```

Reply:
@eric-wieser I managed to switch over to `orbit_rel_zpowers_period`, but it required adding `quotient.mk' q.out'` in a few places in order to avoid the painful issues I was having earlier. Here's the commit (on a separate branch): https://github.com/leanprover-community/mathlib/commit/959778addaa5d3750b128c5b0d4346b70636f329

Reply:
I'm not sure whether this change is worth it, but it would be nice to get this resolved soonish, because this PR is holding up a couple more group theory results (Burnside's transfer theorem and then [G:Z(G)]th power map being a homomorphism).

Reply:
I personally don't see an issue with the use of `.out'` here, because the definition is fundamentally about choosing a representative from each orbit. And the use of `.out'` makes things smooth both in this PR and in the followup PR #14520 (indeed, the commit https://github.com/leanprover-community/mathlib/commit/959778addaa5d3750b128c5b0d4346b70636f329 seems to make the key transfer computation more painful to prove).

Reply:
Ah you're right, the use of `.out'` in `quotient_equiv_sigma_zmod_symm_apply` is unavoidable because the definition really is making a choice; there is no canonical choice of `q.orbit ≃ zmod (orbit_zpowers_period _ _ q)` without picking an initial element.

Reply:
Do you think `orbit_rel_zpowers_period` is worth having anyway, or not?

Reply:
It's good to have in mathlib, but here I think it makes things more clunky. Things seems to work smoother here when everything is in terms of `q.out'`.